rfcs/rdfa-editor-stage-1.md

```diff
@@ -0,0 +1,587 @@
+any piece of code can mutate the (in essence global) document state.
+
+We can learn from ProseMirror that a functional style lends itself well to these requirements. Treating the document as a value means things like
+combining transformations, history rollbacks, and asynchronous editing become simpler to reason about. Edits can become transactions, not unlike database
+transactions, which we like for their atomicity and ability to rollback to a known valid state when things go wrong.
+
+
+### Architecture Stage 1
+
+To give a quick overview of the ideas in this proposal, we will give an example of how a typical editing command would be implemented:
+
+The details and structuring of the various utility methods implied here are not part of this RFC, but they show a 
+philosophy of where various pieces of logic are expected to live.
+
+From the start of the VDOM-period, I've went back and forth about whether node classes
+should contain logic, or wether we should adopt a more functional design where 
+we make more heavy use of utility functions. From our experiments with a ModelTable
+custom node class, we learn that node-owned methods can be very expressive and intuitive.
+The question of "what can I do with this" becomes easily answered. 
+
+The main argument against this usually involved control over state changes. How
+do we implement a `table.insertRow` method while also making sure we can control and track any changes
+to the document? The current implementation of mutators and operations provides a solution, but
+we run into weird situations where we have methods on nodes which all take in a mutator, which they need to get from somewhere.
+It also puts a lot of strain on the mutator's api surface, as different node types can have different abstractions and might
+want to interact with their content in different ways.
+
+These problems dissapear if we apply the immutability principle here. If we demand that any node-owned method
+does not mutate the node itself (rather returns a new node with desired values), we can reduce the mutator api surface
+to convenient ways to replace existing nodes (or ranges) with their newly minted content. Prosemirror shows us that this
+approach can be implemented in a sufficiently performant way. 
+
+This demand is consciously not enforced, as javascript is a notably _very_ mutable language, and _actually_ making things immutable
+comes with a cost. Plugin authors which implement their own node types are operating on the lowest possible level, and 
+are expected to read documentation and adhere to conventions, just as someone assembling their own computer from parts can be expected
+to handle components with care and follow instructions.
+
+I also consider allowing plugins to extend the available node-types a very powerful and expressive tool. Rather than concerning
+oneself with node-structures, one could implement a TodoItem nodetype, let's say, with natural methods such as `setStatus()` and `schedule()`,
```

Comment:
Also interesting for templates and inputs where they can be nodes with special (on top of the model) components like you show on the demo

Reply:
yeah custom nodes come with their own reader/writers
current design doesn't allow for inline components that way, I'd consider that for stage 2 as well